### PR TITLE
BUG: Concatenation of DFs with all NaT columns and TZ-aware ones breaks #12396

### LINKED ISSUE #12396: Concatenation of DFs with all NaT columns and TZ-aware ones breaks
Concatenating DFs that have columns with all NaTs and TZ-aware ones breaks as of 0.17.1:

```
In [1]: import pandas as pd
In [2]: df1 = pd.DataFrame([[pd.NaT], [pd.NaT]])
In [3]: df1
Out[2]:
    0
0   NaT
1   NaT
In [4]: df2 = pd.DataFrame([[pd.Timestamp('2015/01/01', tz='UTC')], [pd.Timestamp('2016/01/01', tz='UTC')]])
In [5]: df2
Out[4]:
    0
0   2015-01-01 00:00:00+00:00
1   2016-01-01 00:00:00+00:00
In [6]: pd.concat([df1, df2])

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-f61a1ab4009e> in <module>()
----> 1 pd.concat([df1, df2])

/.../env/local/lib/python2.7/site-packages/pandas/tools/merge.py in concat(objs, axis, join, join_axes, ignore_index, keys, levels, names, verify_integrity, copy)
    811                        verify_integrity=verify_integrity,
    812                        copy=copy)
--> 813     return op.get_result()
    814 
    815 

/.../env/local/lib/python2.7/site-packages/pandas/tools/merge.py in get_result(self)
    993 
    994             new_data = concatenate_block_managers(
--> 995                 mgrs_indexers, self.new_axes, concat_axis=self.axis, copy=self.copy)
    996             if not self.copy:
    997                 new_data._consolidate_inplace()

/.../env/local/lib/python2.7/site-packages/pandas/core/internals.py in concatenate_block_managers(mgrs_indexers, axes, concat_axis, copy)
   4454                                                 copy=copy),
   4455                          placement=placement)
-> 4456               for placement, join_units in concat_plan]
   4457 
   4458     return BlockManager(blocks, axes)

/.../env/local/lib/python2.7/site-packages/pandas/core/internals.py in concatenate_join_units(join_units, concat_axis, copy)
   4551     to_concat = [ju.get_reindexed_values(empty_dtype=empty_dtype,
   4552                                          upcasted_na=upcasted_na)
-> 4553                  for ju in join_units]
   4554 
   4555     if len(to_concat) == 1:

/.../env/local/lib/python2.7/site-packages/pandas/core/internals.py in get_reindexed_values(self, empty_dtype, upcasted_na)
   4799 
   4800             if self.is_null and not getattr(self.block,'is_categorical',None):
-> 4801                 missing_arr = np.empty(self.shape, dtype=empty_dtype)
   4802                 if np.prod(self.shape):
   4803                     # NumPy 1.6 workaround: this statement gets strange if all

TypeError: data type not understood
```

Possibly related to #11693, #11705 and the  #11456 family. However, this doesn't appear to be caused by the TZ-aware vs. non-TZ aware problems referenced there. 
Versions:

```
In [4]: pd.show_versions()


INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-77-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.17.1
nose: 1.3.6
pip: 6.0.8
setuptools: 12.0.5
Cython: 0.22
numpy: 1.9.2
scipy: 0.15.1
statsmodels: 0.6.1.post1
IPython: 3.1.0
sphinx: 1.3.1
patsy: 0.2.1
dateutil: 2.4.2
pytz: 2015.4
blosc: 1.2.8
bottleneck: 1.0.0
tables: 3.2.0
numexpr: 2.4.3
matplotlib: None
openpyxl: None
xlrd: 0.9.3
xlwt: 0.7.5
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
Jinja2: None
```

### PR DESCRIPTION
**Pull request by multiloc**
- [X] closes #12396
- [X] tests added / passed
- [X] passes `git diff upstream/master | flake8 --diff`
- [X] whatsnew entry
- `concat` to handle tz-aware all NaT/mixed cases
- added unit tests

### CODE REVIEW

File: pandas/core/internals.py [hunk 1]
```diff
@@ -4802,6 +4802,11 @@ def get_reindexed_values(self, empty_dtype, upcasted_na):
 
             if self.is_null and not getattr(self.block, 'is_categorical',
                                             None):
+                if isinstance(empty_dtype, DatetimeTZDtype):
+                    # handle timezone-aware all NaT cases
```
Comment (jreback) [inline pandas/core/internals.py]: see if u can make this if else flow s bit more

**Comment by hcontrast** [comment]
Could do
```
if self.is_null and not getattr(self.block, 'is_categorical',
                                None):
    num_elements = np.prod(self.shape)
    if isinstance(empty_dtype, DatetimeTZDtype):
        # handle timezone-aware all NaT cases
        return DatetimeIndex([fill_value] * num_elements,
                             dtype=empty_dtype)
    else:
        missing_arr = np.empty(self.shape, dtype=empty_dtype)
        if num_elements:
            # ...
            missing_arr.fill(fill_value)
        return missing_arr
```

File: pandas/core/internals.py (hunk 1 above)
Comment (jreback) [inline pandas/core/internals.py]: what i mean is make this an else-if so this cleanly handles the different dtypes (categorical, datetime with tz) and is obvious what is going on

File: pandas/core/internals.py [hunk 2]
```diff
@@ -4802,6 +4802,11 @@ def get_reindexed_values(self, empty_dtype, upcasted_na):
 
             if self.is_null and not getattr(self.block, 'is_categorical',
                                             None):
+                if isinstance(empty_dtype, DatetimeTZDtype):
```
Comment (jreback) [inline pandas/core/internals.py]: you can also blow away the numpy 1.6 workaround comment / code.

File: pandas/core/internals.py [hunk 3]
```diff
@@ -4802,6 +4802,11 @@ def get_reindexed_values(self, empty_dtype, upcasted_na):
...
+                    # handle timezone-aware all NaT cases
+                    return DatetimeIndex([fill_value] * np.prod(self.shape),
+                                         dtype=empty_dtype)
+
```
Comment (jreback) [inline pandas/core/internals.py]: I think we need a method on ``Block`` like ``empty(self, shape)`` to create an empty block filled with the missing value for that type.

**Comment by jreback** [comment]
@hcontrast see what you can do here. This involves cleaning up some code and making it more generic to handle the extension dtypes. lmk if you get stuck.

**Comment by hcontrast** [comment]
Ok. Will have to wait until the weekend though

**Comment by jreback** [comment]
I modified some things in  #12702

pls update/rebase when that's merged

**Comment by hcontrast** [comment]
Rebased on top of master. Having trouble running tests locally though

File: pandas/tools/tests/test_merge.py [hunk 4]
```diff
@@ -2517,6 +2518,113 @@ def test_concat_rename_index(self):
...
+        second = pd.DataFrame([[pd.NaT]])
+
+        result = pd.concat([first, second], 0)
+        assert_frame_equal(result, expect)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]], dtype=dtype)
+
+        result = pd.concat([first, second], 0)
```
Comment (jreback) [inline pandas/tools/tests/test_merge.py]: always write out parameters ``axis=0`` if you are passing them.

File: pandas/tools/tests/test_merge.py [hunk 5]
```diff
@@ -2517,6 +2518,113 @@ def test_concat_rename_index(self):
...
+        assert_frame_equal(result, expect, check_dtype=False)
+        self.assertEqual(result.dtypes[0], dtype)
```
Comment (jreback) [inline pandas/tools/tests/test_merge.py]: we DO want to ``check_dtype``, no need for a separate ``self.assertEqual``

**Comment by jreback** [comment]
sorry, meant here. can you rebase/update?

File: pandas/tools/tests/test_merge.py (hunk 5 above)
Comment (hcontrast) [inline pandas/tools/tests/test_merge.py]: This was a work-around for a concat bug if I remember correctly

Comment (jreback) [inline pandas/tools/tests/test_merge.py]: that's the point here - this or should fix that bug

**Comment by hcontrast** [comment]
Updated the tests, but the mixed tz-/non-tz-aware cases are now breaking again:

```python
======================================================================
ERROR: test_concat_NaT_dataframes_all_NaT_axis_0 (pandas.tools.tests.test_merge.TestConcatenate)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/.../dev/pandas/pandas/tools/tests/test_merge.py", line 2606, in test_concat_NaT_dataframes_all_NaT_axis_0
    result = pd.concat([first, second], axis=0)
  File "/.../dev/pandas/pandas/tools/merge.py", line 836, in concat
    return op.get_result()
  File "/.../dev/pandas/pandas/tools/merge.py", line 1024, in get_result
    concat_axis=self.axis, copy=self.copy)
  File "/.../dev/pandas/pandas/core/internals.py", line 4544, in concatenate_block_managers
    for placement, join_units in concat_plan]
  File "/.../dev/pandas/pandas/core/internals.py", line 4641, in concatenate_join_units
    for ju in join_units]
  File "/.../dev/pandas/pandas/core/internals.py", line 4907, in get_reindexed_values
    missing_arr = np.empty(self.shape, dtype=empty_dtype)
TypeError: data type not understood
```
where
```python
# pandas/tools/tests/test_merge.py:2601

# one side timezone-aware
dtype = DatetimeTZDtype('ns', tz='UTC')
first = pd.DataFrame([[pd.NaT], [pd.NaT]], dtype=dtype)

result = pd.concat([first, second], axis=0)
```
Looks like accidental/eager upcasting on closer inspection
```
block:  DatetimeTZBlock: slice(0, 1, 1), 1 x 2, dtype: datetime64[ns, UTC]
is_datetime_tz:  True
empty dtype:  datetime64[ns, UTC]
block:  DatetimeBlock: slice(0, 1, 1), 1 x 1, dtype: datetime64[ns]
is_datetime_tz:  False
empty dtype:  datetime64[ns, UTC]
```
The second block is regular datetime but gets passed a tz-aware empty value, hence triggering the wrong case.

**Comment by jreback** [comment]
ok I just pushed a change where we moved all of the concat stuff together (now in ``pandas/types/concat.py``, was mostly from ``core/common.py``).

use this to handle the path
 ```
com.is_extension_type(empty_dtype)
True
```

**Comment by hcontrast** [comment]
Thanks. I've applied that fix, only outstanding issue is the mixed Series-DF case, which doesn't preserve the dtype information correctly:

```
FAIL: test_concat_NaT_series_dataframe_all_NaT (pandas.tools.tests.test_merge.TestConcatenate)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/.../dev/pandas/pandas/tools/tests/test_merge.py", line 2696, in test_concat_NaT_series_dataframe_all_NaT
    assert_frame_equal(result, expect, check_dtype=True)
  File "/.../dev/pandas/pandas/util/testing.py", line 1166, in assert_frame_equal
    obj='DataFrame.iloc[:, {0}]'.format(i))
  File "/.../dev/pandas/pandas/util/testing.py", line 1026, in assert_series_equal
    assert_attr_equal('dtype', left, right)
  File "/.../dev/pandas/pandas/util/testing.py", line 804, in assert_attr_equal
    left_attr, right_attr)
  File "/.../dev/pandas/pandas/util/testing.py", line 915, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: Attributes are different

Attribute "dtype" are different
[left]:  object
[right]: datetime64[ns]

```

```python
    def test_concat_NaT_series_dataframe_all_NaT(self):
        # GH 12396

        # non-timezone aware
        first = pd.Series([pd.NaT, pd.NaT])
        second = pd.DataFrame([[pd.Timestamp('2015/01/01')],
                               [pd.Timestamp('2016/01/01')]])

        expect = pd.DataFrame([pd.NaT, pd.NaT, second.iloc[0, 0],
                               second.iloc[1, 0]], index=[0, 1, 0, 1])

        result = pd.concat([first, second])
        assert_frame_equal(result, expect)

        # one side timezone-aware
        dtype = DatetimeTZDtype('ns', tz='UTC')
        first = first.astype(dtype)
        result = pd.concat([first, second])
        assert_frame_equal(result, expect, check_dtype=True)
        self.assertEqual(result.dtypes.iloc[0], second.dtypes[0])

        # both sides timezone-aware
        second = second.apply(lambda x: x.astype(dtype))

        result = pd.concat([first, second])
        assert_frame_equal(result, expect, check_dtype=True)
        self.assertEqual(result.dtypes.iloc[0], second.dtypes[0])
```

**Comment by jreback** [comment]
that last already seems to work

**Comment by hcontrast** [comment]
Might be due to the patch then

**Comment by jreback** [comment]
can you rebase / update

**Comment by hcontrast** [comment]
updated

File: pandas/core/internals.py [hunk 6]
```diff
@@ -4911,6 +4911,11 @@ def get_reindexed_values(self, empty_dtype, upcasted_na):
                     pass
                 elif getattr(self.block, 'is_sparse', False):
                     pass
+                elif com.is_extension_type(empty_dtype):
+                    num_elements = np.prod(self.shape)
+                    # handle timezone-aware all NaT cases
```
Comment (jreback) [inline pandas/core/internals.py]: just do this if it's a datetime tz type

File: pandas/core/internals.py [hunk 7]
```diff
@@ -4911,6 +4911,12 @@ def get_reindexed_values(self, empty_dtype, upcasted_na):
                     pass
                 elif getattr(self.block, 'is_sparse', False):
                     pass
+                elif com.is_extension_type(empty_dtype) and \
+                        com.is_datetimetz(empty_dtype):
+                    num_elements = np.prod(self.shape)
+                    # handle timezone-aware all NaT cases
```
Comment (jreback) [inline pandas/core/internals.py]: no, do this under the ``if getattr(self.block, 'is_datetimetz', False) or com.is_datetimetz(empty_dtype)``

Comment (hcontrast) [inline pandas/core/internals.py]: Like so? This passes the tests but not sure how good the coverage is for the non-NaT tz-aware cases.
```python
                if getattr(self.block, 'is_datetimetz', False) \
                    or com.is_datetimetz(empty_dtype):
                    num_elements = np.prod(self.shape)
                    # handle timezone-aware all NaT cases
                    return DatetimeIndex([fill_value] * num_elements,
                                         dtype=empty_dtype)
                elif getattr(self.block, 'is_categorical', False):
                    pass
                elif getattr(self.block, 'is_sparse', False):
                    pass
```

Comment (jreback) [inline pandas/core/internals.py]: yes more like this. yes coverage of extension types is prob only so now

File: pandas/tools/tests/test_merge.py [hunk 8]
```diff
@@ -2522,6 +2523,120 @@ def test_concat_multiindex_with_tz(self):
...
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expect)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]], dtype=dtype)
+
+        result = pd.concat([first, second], axis=0)
+        # upcasts for mixed case
+        assert_frame_equal(result, expect, check_dtype=False)
```
Comment (jreback) [inline pandas/tools/tests/test_merge.py]: make a new expect here that matches exactly what you want rather, so ``assert_frame_equal(result, expected)`` works & checks dtypes

File: pandas/tools/tests/test_merge.py [hunk 9]
```diff
@@ -2522,6 +2523,120 @@ def test_concat_multiindex_with_tz(self):
...
+        self.assertEqual(result.dtypes[0], dtype)
+
+        # both sides timezone-aware
+        second = pd.DataFrame([[pd.NaT]], dtype=dtype)
+
+        result = pd.concat([first, second], axis=0)
+        # upcasts to tz-aware
+        assert_frame_equal(result, expect, check_dtype=False)
+        self.assertEqual(result.dtypes[0], dtype)
```
Comment (jreback) [inline pandas/tools/tests/test_merge.py]: same here

File: pandas/tools/tests/test_merge.py [hunk 10]
```diff
@@ -2522,6 +2523,120 @@ def test_concat_multiindex_with_tz(self):
...
+
+        result = pd.concat([first, second], axis=1)
+        assert_frame_equal(result, expect)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]], dtype=dtype)
+
+        # upcasts result to tz-aware
+        assert_frame_equal(result, expect, check_dtype=False)
```
Comment (jreback) [inline pandas/tools/tests/test_merge.py]: same here, make an exactly frame of what you are comparing.

File: pandas/tools/tests/test_merge.py [hunk 11]
```diff
@@ -2522,6 +2523,120 @@ def test_concat_multiindex_with_tz(self):
...
+        self.assertEqual(result.dtypes.iloc[1], second.dtypes[0])
+
+        # both sides timezone-aware
+        second = pd.DataFrame([[pd.NaT]], dtype=dtype)
+
+        result = pd.concat([first, second], axis=1)
+        assert_frame_equal(result, expect, check_dtype=False)
+        # upcasts to tz-aware
+        self.assertEqual(result.dtypes.iloc[0], dtype)
+        self.assertEqual(result.dtypes.iloc[0], first.dtypes[0])
```
Comment (jreback) [inline pandas/tools/tests/test_merge.py]: same

**Comment by jreback** [comment]
ok (this is the original issue)
```
In [1]: df1 = pd.DataFrame([[pd.NaT], [pd.NaT]])

In [2]: df2 = pd.DataFrame([[pd.Timestamp('2015/01/01', tz='UTC')], [pd.Timestamp('2016/01/01', tz='UTC')]])

In [4]: pd.concat([df1, df2])
Out[4]: 
                          0
0                       NaT
1                       NaT
0 2015-01-01 00:00:00+00:00
1 2016-01-01 00:00:00+00:00

In [5]: pd.concat([df1, df2]).dtypes
Out[5]: 
0    datetime64[ns, UTC]
dtype: object
```

This is incorrect; we are coercing to object (correct), but the ``NaT`` are getting incorrectly corced to ``nan`` here.
```
In [6]: df2 = pd.DataFrame([[pd.Timestamp('2015/01/01', tz='UTC')], [pd.Timestamp('2016/01/01', tz='US/Eastern')]])

In [7]: pd.concat([df1, df2]).dtypes
Out[7]: 
0    object
dtype: object

In [8]: pd.concat([df1, df2])
Out[8]: 
                           0
0                        NaN
1                        NaN
0  2015-01-01 00:00:00+00:00
1  2016-01-01 00:00:00-05:00
```

**Comment by jreback** [comment]
@hcontrast can you update. only small fix I think.

**Comment by hcontrast** [comment]
Added test case for this, indeed still breaking

**Comment by hcontrast** [comment]
Can't actually get the test to break. This should break but doesn't

```python
        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
        second = pd.DataFrame([[pd.Timestamp('2015/01/01', tz='UTC')],
                               [pd.Timestamp('2016/01/01', tz='US/Eastern')]])

        expect = pd.DataFrame([pd.NaT, pd.NaT, second.iloc[0, 0],
                               second.iloc[1, 0]], index=[0, 1, 0, 1])

        result = pd.concat([first, second], axis=0)
        assert_frame_equal(result, expect)
        print expect
        print result
```
gives
```
                           0
0                        NaT
1                        NaT
0  2015-01-01 00:00:00+00:00
1  2016-01-01 00:00:00-05:00
                           0
0                        NaN
1                        NaN
0  2015-01-01 00:00:00+00:00
1  2016-01-01 00:00:00-05:00
```

**Comment by jreback** [comment]
This is a tricky one; we have this flag to test exactness (so do this for now).

```
In [27]: tm.assert_frame_equal(result, expect, check_datetimelike_compat=True)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-27-441a8feab35d> in <module>()
----> 1 tm.assert_frame_equal(result, expect, check_datetimelike_compat=True)

/Users/jreback/pandas/pandas/util/testing.pyc in assert_frame_equal(left, right, check_dtype, check_index_type, check_column_type, check_frame_type, check_less_precise, check_names, by_blocks, check_exact, check_datetimelike_compat, check_categorical, check_like, obj)
   1268                 check_datetimelike_compat=check_datetimelike_compat,
   1269                 check_categorical=check_categorical,
-> 1270                 obj='DataFrame.iloc[:, {0}]'.format(i))
   1271 
   1272 

/Users/jreback/pandas/pandas/util/testing.pyc in assert_series_equal(left, right, check_dtype, check_index_type, check_series_type, check_less_precise, check_names, check_exact, check_datetimelike_compat, check_categorical, obj)
   1131         else:
   1132             assert_numpy_array_equal(left.get_values(), right.get_values(),
-> 1133                                      check_dtype=check_dtype)
   1134     else:
   1135         _testing.assert_almost_equal(left.get_values(), right.get_values(),

/Users/jreback/pandas/pandas/util/testing.pyc in assert_numpy_array_equal(left, right, strict_nan, check_dtype, err_msg, obj)
   1034     # compare shape and values
   1035     if not array_equivalent(left, right, strict_nan=strict_nan):
-> 1036         _raise(left, right, err_msg)
   1037 
   1038     if check_dtype:

/Users/jreback/pandas/pandas/util/testing.pyc in _raise(left, right, err_msg)
   1028             msg = '{0} values are different ({1} %)'\
   1029                 .format(obj, np.round(diff, 5))
-> 1030             raise_assert_detail(obj, msg, left, right)
   1031 
   1032         raise AssertionError(err_msg)

/Users/jreback/pandas/pandas/util/testing.pyc in raise_assert_detail(obj, message, left, right)
    983 [left]:  {2}
    984 [right]: {3}""".format(obj, message, left, right)
--> 985     raise AssertionError(msg)
    986 
    987 

AssertionError: numpy array are different

numpy array values are different (50.0 %)
[left]:  [nan, nan, 2015-01-01 00:00:00+00:00, 2016-01-01 00:00:00-05:00]
[right]: [NaT, NaT, 2015-01-01 00:00:00+00:00, 2016-01-01 00:00:00-05:00]
```

@sinhrks you think we should be checking for this by default? (I forgot why I added this, IIRC maybe for stata tests or something)

**Comment by sinhrks** [comment]
Or we should change ``assert_numpy_array_equal`` to use ``strict_nan=True``. Not sure how many tests are relying on this.

**Comment by jreback** [comment]
@hcontrast can you update

**Comment by jreback** [comment]
can you rebase and update

**Comment by jreback** [comment]
can you rebase & update

**Comment by jreback** [comment]
can you rebase / update?

**Comment by jreback** [comment]
closing, but pls reopen / comment if you can continue

**Comment by jreback** [comment]
this is almost there .....

**Comment by paul-mannino** [comment]
Want to knock this one out in conjunction with #18447. Just one question, what's the general desired behavior for missing data in concatenations? Should NaTs and nans always remain as they are?

**Comment by jreback** [comment]
nan is always by type
eg NaT for daetimelile and nan otherwise

the tests look pretty good here

**Comment by jreback** [comment]
go ahead and open a new PR

**Pull request closed without merging**